core/src/test/java/io/grpc/internal/DelayedStreamTest.java

```diff
@@ -154,6 +181,82 @@ public void setStream_flowControl() {
     verify(realStream).request(3);
   }
 
+  @Test
+  public void setStreamThenStart() {
+    stream.optimizeForDirectExecutor();
+    stream.setCompressor(mock(Compressor.class));
+    stream.setFullStreamDecompression(false);
+    stream.setDecompressorRegistry(DecompressorRegistry.emptyInstance());
+    stream.setDeadline(Deadline.after(1, TimeUnit.MINUTES));
+    stream.setAuthority("auth");
+    stream.setMaxInboundMessageSize(10);
+    stream.setMaxOutboundMessageSize(10);
+
+    // not started yet, nothing to drain
+    assertNull(stream.setStream(realStream));
+    verifyNoInteractions(realStream);
+    assertEquals(8, stream.getPreStartPendingCallsCount());
```

Comment:
Instead of adding these white-box methods, can you instead have the test verify in a black-box style? Basically, the `verifyNoInteractions(realStream)` is the appropriate white-box verification here. After calling `start()` you'd verify that the various pre-start methods had been called.

Reply:
Basically, it is better to test the API instead of internal behaviors, as the API is more stable to changes and reliable in verifying the expected behavior. Sometimes we do need methods that peek inside the implementation, but it doesn't seem necessary here.

Reply:
Sounds fair. Updated. It might overlap with existing ones but more stress way so put it in a separate test case.